# Citations

## [RDscan](https://doi.org/10.1128/mSphere.00535-21)

> D. Bespiatykh, J. Bespyatykh, I. Mokrousov, and E. Shitikov, A Comprehensive Map of Mycobacterium tuberculosis Complex Regions of Difference, mSphere, Volume 6, Issue 4, 21 July 2021, Page e00535-21, https://doi.org/10.1128/mSphere.00535-21

## Pipeline and package menagers

- [Sankemake](https://f1000research.com/articles/10-33/v1/)

  > Mölder F, Jablonski KP, Letcher B et al. Sustainable data analysis with Snakemake [version 1; peer review: 1 approved, 1 approved with reservations]. F1000Research 2021, 10:33.

- [Conda](https://docs.conda.io/en/latest/)

- [BioConda](https://www.nature.com/articles/s41592-018-0046-7#citeas/)
  > Grüning, B., Dale, R., Sjödin, A. et al. Bioconda: sustainable and comprehensive software distribution for the life sciences. Nat Methods 15, 475–476 (2018). https://doi.org/10.1038/s41592-018-0046-7

## Pipeline tools

- [BWA-MEM](https://arxiv.org/abs/1303.3997v2/)

  > Li H. (2013) Aligning sequence reads, clone sequences and assembly contigs with BWA-MEM. arXiv:1303.3997v2.

- [SAMtools](https://academic.oup.com/bioinformatics/article/27/21/2987/217423/)

  > Heng Li, A statistical framework for SNP calling, mutation discovery, association mapping and population genetical parameter estimation from sequencing data, Bioinformatics, Volume 27, Issue 21, 1 November 2011, Pages 2987–2993, https://doi.org/10.1093/bioinformatics/btr509

- [BEDTools](https://academic.oup.com/bioinformatics/article/26/6/841/244688/)

  > Aaron R. Quinlan, Ira M. Hall, BEDTools: a flexible suite of utilities for comparing genomic features, Bioinformatics, Volume 26, Issue 6, 15 March 2010, Pages 841–842, https://doi.org/10.1093/bioinformatics/btq033

- [SURVIVOR](https://www.nature.com/articles/ncomms14061/)

  > Jeffares, D., Jolly, C., Hoti, M. et al. Transient structural variations have strong effects on quantitative traits and reproductive isolation in fission yeast. Nat Commun 8, 14061 (2017). https://doi.org/10.1038/ncomms14061

- [GATK](https://www.nature.com/articles/ng.806/)

  > DePristo, M., Banks, E., Poplin, R. et al. A framework for variation discovery and genotyping using next-generation DNA sequencing data. Nat Genet 43, 491–498 (2011). https://doi.org/10.1038/ng.806

- [BCFtools](https://academic.oup.com/gigascience/article/10/2/giab008/6137722/)

  > Petr Danecek, James K Bonfield, Jennifer Liddle, John Marshall, Valeriu Ohan, Martin O Pollard, Andrew Whitwham, Thomas Keane, Shane A McCarthy, Robert M Davies, Heng Li, Twelve years of SAMtools and BCFtools, GigaScience, Volume 10, Issue 2, February 2021, giab008, https://doi.org/10.1093/gigascience/giab008

- [Duphold](https://academic.oup.com/gigascience/article/8/4/giz040/5477467/)

  > Brent S Pedersen, Aaron R Quinlan, Duphold: scalable, depth-based annotation and curation of high-confidence structural variant calls, GigaScience, Volume 8, Issue 4, April 2019, giz040, https://doi.org/10.1093/gigascience/giz040

- [SnpEff](https://www.tandfonline.com/doi/full/10.4161/fly.19695/)

  > Pablo Cingolani, Adrian Platts, Le Lily Wang, Melissa Coon, Tung Nguyen, Luan Wang, Susan J. Land, Xiangyi Lu & Douglas M. Ruden (2012) A program for annotating and predicting the effects of single nucleotide polymorphisms, SnpEff, Fly, 6:2, 80-92, DOI: 10.4161/fly.19695

- [Mosdepth](https://academic.oup.com/bioinformatics/article/34/5/867/4583630/)

  > Brent S Pedersen, Aaron R Quinlan, Mosdepth: quick coverage calculation for genomes and exomes, Bioinformatics, Volume 34, Issue 5, 01 March 2018, Pages 867–868, https://doi.org/10.1093/bioinformatics/btx699

- [R](https://www.r-project.org/)
  > R Core Team (2021). R: A language and environment for statistical computing. R Foundation for Statistical Computing, Vienna, Austria. URL https://www.R-project.org/
